### DISCUSSION TITLE
Understanding how to batch multiple graphs with the same number of nodes.

### DISCUSSION THREAD

**Discussion by AlexNiko47** [pyg-team/pytorch_geometric#5434] [Q&A] [answered]
Hi,

I'm trying to implement a graph classification model and I'm struggling to understand how to properly batch a list of graphs and feed them into the model. All my graphs are stored in a list and are of the form `Data(x=[978], edge_index=[2, 956484], edge_attr=[956484], y=[1, 38])` and I'm first trying to implement a one layer linear model to test whether my approach works. Here's the code:

`Loader = DataLoader(list_of_graphs,batch_size=2, shuffle=True)`
`LinearModel(torch.nn.Module):
    def __init__(self, in_channels: int, out_channels: int):
        super().__init__()
        self.linear = Linear(in_channels,out_channels)
    def forward(self,x):
        x = self.linear(x)
        out = torch.sigmoid(x)
        return out `

`def train(model, optimizer, train_loader, device, loss):
    model.train()
    for data in train_loader:
        optimizer.zero_grad()
        data = data.to(device)
        out = model(data.x)
        loss_tensor = loss(out, data.y.to(torch.float))
        print("Loss is: ", loss_tensor.item())
        loss_tensor.backward()
        optimizer.step()
`

`model = LinearModel(978,38)
optimizer = torch.optim.Adam(model.parameters(), lr=0.005, weight_decay=0)
loss = torch.nn.MSELoss()
train(model,optimizer,Loader,"cpu", loss)`

For some reason whenever I try to run this I get the error `RuntimeError: mat1 and mat2 shapes cannot be multiplied (1x1956 and 978x38)` with 1956 being I'm guessing the two graphs from the batch. What is it I'm donig wrong?

**Comment by EdisonLeeeee** [comment, marked as answer]
Hi @AlexNiko47 

In `Data`, `x` must have at least 2 dims, indicating `[num_nodes, num_features]`. A `DataLoader` would stack the node feature `x` at the first dimension from each batch by default. So you got 1x1956 as input with batch size 2.

Now I have a question, does 978 denote the feature dimension or not? If it does, there must something wrong with your data as it has only one node in each batch.

**Reply by AlexNiko47** [reply]
Hi @EdisonLeeeee 
I have a gene expression dataset, where there are 978 genes so I would want to have a `[978,1]` dimension matrix with each gene being a node with 1 feature (the associated expression value for that gene). I create these `Data` objects from a pandas DataFrame like this:
`list_of_graphs = []
for i in training.index:
    x = training.iloc[i,:].to_numpy()
    #x = x[ :, np.newaxis]
    y = labels.iloc[i,:].values
    y = y[np.newaxis, :]
    graph = Data(x = torch.as_tensor(x, dtype=torch.float32), edge_index=coo[0], edge_attr=coo[1],y = torch.as_tensor(y, dtype= torch.float32))
    list_of_graphs.append(graph)`

From the DataFrame I convert each row corresponding to a single experiment to a `(978,)` np array. If I add an additional dimension with np.newaxis and make `x` of shape `[978,1]` I get the same error but now with the node features stacked in the first dimension: `RuntimeError: mat1 and mat2 shapes cannot be multiplied (1956x1 and 978x38)` so maybe the way in which I create the DataLoader object might be the problem?

**Reply by EdisonLeeeee** [reply]
I think you have to add an additional dimension for `x` and make it of shape `[978, 1]`. Correspondingly, the `in_channels` of your model should be `1` instead of `978`.

**Reply by AlexNiko47** [reply]
Hi @EdisonLeeeee 
I've just tried this and now I'm getting the following error: `RuntimeError: The size of tensor a (1956) must match the size of tensor b (2) at non-singleton dimension 0`

The shape of the `data.x` is `torch.Size([1956, 1])` and for `data.y` is `torch.Size([2, 38])`

**Reply by rusty1s** [reply]
If you want to use `MSELoss`, the shapes of the prediction and targets need to be the same. As such, it looks like you are missing a global pooling function such as `torch_geometric.nn.global_mean_pool` to convert node-level features into graph-level ones.

**Reply by AlexNiko47** [reply]
The pooling layer seems to have done the trick! Thanks a lot :)
